services/kill.py: Allow admins to kill any run

VallariAg opened the pull request:
Admins are users who are on "Ceph" team on github. 
This PR also improves error handling for kill route.  


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [Issue/Tracker URL]
      Signed-off-by: [Your Name] <[your email]>

-->


## Contribution Guidelines

To sign and test your commits, please refer to [Contibution guidelines](./../CONTRIBUTING.md).

## Checklist
- [x] Changes tested in [container setup](./../README.md#option-1-teuthology-docker-setup)
- [ ] Changes tested in [non-containerized setup](./../README.md#option-2-non-containerized-with-venv-and-pip)
- [x] Includes tests
- [ ] Documentation updated

VallariAg committed:
services/kill.py: Allow admins to kill any run

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg committed:
services/kill.py: Owner can also be "scheduled_<username>@teuthology"

"scheduled_<username>@teuthology" is the default owner name if
run is scheduled from teuthology CLI tool. This commit allows
users of same github username to recognize it as their jobs.

Also fix how run_owner is determined. In teuthology, runs ownership
is deteremined by job's "owner" and not run's "user" value.

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg committed:
login: add 'isUserAdmin' to session data

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg committed:
kill: add tests + fix run owner logic

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg committed:
Improve error handling

And decode kill-cmd output logs.

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg committed:
services/kill.py: optionally add boolean value flag in kill cmd

For example, value of "--preserve-queues" is boolean in /kill request
schema. If it's true, add it to the kill cmd, otherwise don't add the
flag.

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

kamoltat commented on src/teuthology_api/services/helpers.py:
```diff
@@ -96,3 +99,19 @@ def get_token(request: Request):
+
+async def isAdmin(username, token):
+    TEAM_MEMBER_URL = (
+        f"https://api.github.com/orgs/ceph/teams/{ADMIN_TEAM}/memberships/{username}"
```
Maybe this piece: https://api.github.com/orgs/ceph/teams/ can be an env var?

kamoltat commented on src/teuthology_api/services/helpers.py:
```diff
@@ -96,3 +99,19 @@ def get_token(request: Request):
+        }
+        response_org = await client.get(url=TEAM_MEMBER_URL, headers=headers)
+        response_org_dic = dict(response_org.json())
+        if response_org_dic.get("state") == "active":
```
if "state" doesn't exists then is there a chance response_org_dic = [ ] ? and will somehow stop the program?

VallariAg replied:
I did a quick check and doesn't look like it would throw an error:
```
>>> dict([])
{}
```
But I can add more error handling, thanks!

kamoltat commented on src/teuthology_api/services/kill.py:
```diff
@@ -26,38 +28,48 @@ def run(args, send_logs: bool, access_token: str, request: Request):
     if run_name:
         run_details = get_run_details(run_name)
-        run_owner = run_details.get("user")
+        jobs_details = run_details.get("jobs", [])
```
I saw this line and compared it with the line above `if response_org_dic.get("state") == "active":`

We could have some consistency? unless there is a good reason why some `get` you include a default argument

kamoltat reviewed (changes requested):
Minor coding style comments and env variable stuff, other than that, everything else looks fine

kamoltat commented:
<img width="1281" alt="image" src="https://github.com/user-attachments/assets/4d78fcb0-b156-4d90-b2f7-cbb3ed7d0458">

VallariAg committed:
services/helpers.py: handle empty GH_ORG_TEAM_URL and ADMIN_TEAM

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg committed:
main.py: add CORSMiddleware for non-dev env

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

VallariAg committed:
Send user avatar in session data

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

kamoltat commented on src/teuthology_api/services/kill.py:
```diff
@@ -4,60 +4,72 @@
+    if (run_owner.lower() != username.lower()) and (
+        run_owner.lower() != f"scheduled_{username.lower()}@teuthology"
+    ):
+        isUserAdmin = await isAdmin(username, token["access_token"])
```
just a little nit, if token dic exists but the "access_token" key doesn't exist, are we handling this properly?

kamoltat reviewed (dismissed):
Just a little nit, otherwise I think everything else looks great

VallariAg committed:
Handle missing token in isAdmin() func

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

kamoltat commented on src/teuthology_api/routes/kill.py:
```diff
@@ -28,5 +29,16 @@ def create_run(
+    try:
+        args = args.model_dump(by_alias=True, exclude_unset=True)
+        return await run(args, logs, token, request)
+    except HTTPException:
```
should we log the exception here as well? just to keep it consistence with lines below

kamoltat reviewed (dismissed):
minor comment, other than that it is good to go!

VallariAg committed:
routes/kill.py: log error for HTTPException

Errors are logged for HTTPError and Exception,
also log them for HTTPException before raising
the exception.

Signed-off-by: Vallari Agrawal <val.agl002@gmail.com>

kamoltat reviewed (approved):
LGTM
